### PR TITLE
[OTX] Add Self-SL for semantic segmentation

### PR DESCRIPTION
**Pull request by sungchul1**
# This PR includes
* Self-SL for semantic segmentation recipe
* Disenable validation during self-sl training
* Update docstring & variable names in classification

## FYI.
* This PR has dependency with https://github.com/openvinotoolkit/model_preparation_algorithm/pull/122.
* ~~Target branch will be changed to `feature/otx` after merging https://github.com/openvinotoolkit/training_extensions/pull/1457.~~
* ~~[link](https://github.com/openvinotoolkit/training_extensions/pull/1462/files#diff-a79b5906fd5e84faede851a58b8f03f86ac13e192375c9fdad50c1b3b95fb02cR176-R182) was temporarily added to use formats of pasval voc and cityscapes. It will be removed after updating implementation of setting dataset suffix or datumaro integration.~~

## TODOs
- [x] component to create pseudo masks using images
  - ([otx/algorithms/segmentation/adapters/mmseg/utils/data_utils.py L159-199](https://github.com/openvinotoolkit/training_extensions/pull/1462/files#diff-a79b5906fd5e84faede851a58b8f03f86ac13e192375c9fdad50c1b3b95fb02cR159-R199))  
  DetCon uses pseudo masks to get contrastive loss and it is required to make the masks if they don't exist.
- [x] integration test
- [ ] ~~implementation of setting dataset suffix (It can be unnecessary because of datumaro)~~
- [x] remove temp code
- [x] refactoring

## TBD.
* Automatically adding `background` class
> DetCon uses pseudo masks, but there is no relationship between labels in these masks. So I think `meta.json` can be temporarily created like `{"labels_map": [{"name": "target1", "id": 1}, {"name": "target2", "id": 2}, ...]}`. But, otx automatically adds `background` class and `meta.json` should be created with 1 sample less than the number of labels by considering automatically added `background` class. I'd like to suggest to create an option to add `background` class whenever it is needed.

* Issue about automatically extended channel when disenabling to add `background` class
> When disenabling to add `background` class, creating `meta.json` with the number of all labels, and pretraining, an error sometimes occurs like below. I don't know the reason but we don't have to consider this issue in this PR.
```script
...
  File "otx/algorithms/venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "otx/algorithms/venv/lib/python3.8/site-packages/mmseg/models/losses/base.py", line 97, in forward
    loss, meta = self._forward(*args, **kwargs)
  File "otx/algorithms/venv/lib/python3.8/site-packages/mpa/modules/models/losses/mpa_pixel_base.py", line 32, in _forward
    losses, updated_output = self._calculate(output, _labels, valid_label_mask, self._last_scale)
  File "otx/algorithms/venv/lib/python3.8/site-packages/mpa/modules/models/losses/cross_entropy_loss_with_ignore.py", line 59, in _calculate
    loss = F.nll_loss(each_prob_log, each_label, reduction='none', ignore_index=self.ignore_index)
  File "otx/algorithms/venv/lib/python3.8/site-packages/nncf/torch/dynamic_graph/wrappers.py", line 89, in wrapped
    op1 = operator(*args, **kwargs)
  File "otx/algorithms/venv/lib/python3.8/site-packages/torch/nn/functional.py", line 2390, in nll_loss
    ret = torch._C._nn.nll_loss2d(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
IndexError: Target 22 is out of bounds.
```

## Tests
- `tests/integration/cli/segmentation/test_segmentation.py`
image

### CODE REVIEW

File: otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py
Comment (wonjuleee) [inline otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py]: Typo. disable was correct.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (wonjuleee) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: can't we reuse `same_obj_aa` and `same_obj_bb` for here?

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (wonjuleee) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: could you explain why we need to multiply by 0?

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (wonjuleee) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: I think too many variables here from cross view operations.
Can we reduce them by `labels_0 =torch.cat([labels_ab, labels_aa], dim=2).reshape((batch_size, num_samples, -1))`

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (wonjuleee) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: If so, when num_positives_0 is 0, loss_a should be distorted.
If we compute by `labels_0 = labels_0 / (num_positives_0 + epsilon)`, where epsilon = 1e-6, is it expected to be an error?

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (wonjuleee) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py:86]: I believe there must be some smarter ways to express them.
To do this, we can do below as 
for i, (v0, v1) in enumerate(list(itertools.product(view0, view1))):
    same_obj[i] = make_same_obj(v0, v1)

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: `same_obj_aa` is for `pind1` and `tind1`, and `same_obj_bb` is for `pind2` and `tind2`. ([link](https://github.com/openvinotoolkit/training_extensions/blob/feature/otx-warmstart-seg/otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py#L69-L72))
So, `make_same_obj(pind1, pind1)` must be newly set to get `obj_area_0`, and so `make_same_obj(pind2, pind2)`.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: DetCon doesn't consider features from the same view.
But, despite the same view, there must be negative samples corresponding to different masks, so we can exclude positive samples by subtracting them with very large values like L117-118 and consider both `labels_aa` and `labels_bb` as negative samples by multiplying zero.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: I agree with too many variables. But I don't know if combining the two lines is a better option.
We can reduce the source line from 2 to 1, but the reduced line will be long and complicated.
I don't know which approach is the best, so we need to discuss it.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py]: To overcome zero division error, `torch.maximum` was used.
We can use 1 instead when `num_positives_0` is zero.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py:86]: I agree there are so many parts to be refactored.
First of all, I changed the official code from jax to pytorch and left it as it is.
I'm not sure which way is more efficient and intuitive.
But there are some parts to be refactored, such as pipelines (currently: np -> PIL -> np, it is required to be changed.) and modules (for instance, `SelfSLMLP` is used for both cls and seg, and it is necessary to combine them.), so DetCon loss function simplification can be processed with the above process together.

File: otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py
Comment (sungchul1) [reply otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py]: If functional changes are required, I'll update it with that process.
If not, it can be processed with the refactoring to be processed later together.

**Comment by sungchul1** [comment]
I checked [previous pre-merge test logs](https://github.com/openvinotoolkit/training_extensions/actions/runs/3787394710/jobs/6439228052) and `TestToolsMPAMultilabelClassification::test_otx_eval_openvino` was only failed.
I think this PR can be merged if functional change is not required.
```shell
2022-12-27T16:39:03.0914637Z FAILED tests/integration/cli/classification/test_classification.py::TestToolsMPAMultilabelClassification::test_otx_eval_openvino[Custom_Image_Classification_EfficientNet-V2-S]
```

**Comment by sungmanc** [comment]
> I checked [previous pre-merge test logs](https://github.com/openvinotoolkit/training_extensions/actions/runs/3787394710/jobs/6439228052) and `TestToolsMPAMultilabelClassification::test_otx_eval_openvino` was only failed. I think this PR can be merged if functional change is not required.
> 
> ```shell
> 2022-12-27T16:39:03.0914637Z FAILED tests/integration/cli/classification/test_classification.py::TestToolsMPAMultilabelClassification::test_otx_eval_openvino[Custom_Image_Classification_EfficientNet-V2-S]
> ```

Before merging, it is needed to fix the pre-commit FAIL case. resolving all comments also is needed.

**Comment by JihwanEom** [comment]
Could you rebase and rerun tests? Multi-label bugfix PR is merged today morning.  https://github.com/openvinotoolkit/training_extensions/pull/1436

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py:86]: I agree with Wonju. I hope you can reduce the duplicated operations here.

File: otx/algorithms/segmentation/configs/base/data/selfsl/data_pipeline.py
Comment (supersoob) [inline otx/algorithms/segmentation/configs/base/data/selfsl/data_pipeline.py]: I found that from ColorJitter to GaussianBlur each pipeline includes the steps to convert to PIL images. I think you can fix this conversion at once by putting [such pipeline methods](https://github.com/openvinotoolkit/model_preparation_algorithm/blob/8c17768043def5e7ef9023fcb5b9cc9b3cb3a987/mpa/modules/datasets/pipelines/torchvision2mmdet.py#L151) both in the start and end of pipelines. And I also wanna suggest to include probs arguments for the pipeline method itself rather than using ProbCompose to remove redundant format conversion.

File: otx/algorithms/segmentation/adapters/mmseg/models/necks/selfsl_mlp.py
Comment (supersoob) [inline otx/algorithms/segmentation/adapters/mmseg/models/necks/selfsl_mlp.py:19]: Is this neck class exactly same with the one in classification? Actually I don't really think that reusing is always better than copying. And as far as I know, it is plan to support the all the model replacement not only for backbone in the future. I think it's okay to leave it as is. But if needed, I suggest to import the registry function dynamically checking the task type argument when this neck class is called initially.

**Review by supersoob — COMMENTED** [review]
I just left few comments. Are you going to refactor the codes here or other PR.

File: otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py
Comment (sungchul1) [reply otx/algorithms/classification/adapters/mmcls/models/classifiers/byol.py]: Done.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py:86]: Done.

File: otx/algorithms/segmentation/configs/base/data/selfsl/data_pipeline.py
Comment (sungchul1) [reply otx/algorithms/segmentation/configs/base/data/selfsl/data_pipeline.py]: Thanks for the recommendation. I added that conversion pipeline to reduce redundant conversion and added probability in each module to avoid using `ProbCompose`, and updated the recipe.

File: otx/algorithms/segmentation/adapters/mmseg/models/necks/selfsl_mlp.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/models/necks/selfsl_mlp.py:19]: Exactly same. I'd like to know the reason why you don't think that reusing is always better than copying.
And following the plan you said, I'll update this module later if it is necessary to combine them.

**Comment by sungchul1** [comment]
Local test result (semantic segmentation)
image

File: otx/algorithms/classification/tasks/inference.py
Comment (harimkang) [inline otx/algorithms/classification/tasks/inference.py:336]: The name is changed once more here. @JihwanEom It seems necessary to refer to it.

File: otx/algorithms/segmentation/adapters/mmseg/utils/data_utils.py
Comment (harimkang) [inline otx/algorithms/segmentation/adapters/mmseg/utils/data_utils.py:210]: It seems that this will only work under very specific conditions. Also, I think it may not be used after datumaro integration. Do you have any plans to fix this?

File: otx/algorithms/segmentation/tasks/inference.py
Comment (harimkang) [inline otx/algorithms/segmentation/tasks/inference.py]: It is minor, but since the number of tasks has increased, how about making (TrainType.SEMISUPERVISED, TrainType.INCREMENTAL, TrainType.SELFSUPERVISED) a global variable for the file?

refer: https://github.com/openvinotoolkit/training_extensions/blob/c076902a791e069cb89abab89550cc13449d6b53/otx/cli/tools/find.py#L30

**Review by harimkang — COMMENTED** [review]
I left a few minor comments.

File: otx/algorithms/segmentation/adapters/mmseg/utils/data_utils.py
Comment (sungchul1) [reply otx/algorithms/segmentation/adapters/mmseg/utils/data_utils.py:210]: I agree. There are also other temp conditions to be fixed after the main functional integrations, such as datumaro, in this PR.
I will fix them with another PR.

File: otx/algorithms/segmentation/tasks/inference.py
Comment (sungchul1) [reply otx/algorithms/segmentation/tasks/inference.py]: Done.

**Review by supersoob — DISMISSED** [review]

**Review by harimkang — DISMISSED** [review]
LGTM! Let's wait for the CI test results!

**Review by supersoob — APPROVED** [review]

**Review by harimkang — APPROVED** [review]
LGTM! If there is no problem with the CI test, I think it will be okay to merge.

**Comment by sungchul1** [comment]
There are 5 failed cases in pre-merge test. It seems that they are not related to Self-SL for semantic segmentation.

- tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_otx_demo[Custom_Image_Classification_EfficientNet-V2-S]
`RuntimeError: CUDA error: out of memory`

- tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficientNet-V2-S]
`RuntimeError: CUDA error: out of memory`

- tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficinetNet-B0]
`RuntimeError: CUDA error: out of memory`

- tests/integration/cli/detection/test_detection.py::TestToolsMPADetection::test_otx_explain[Custom_Object_Detection_YOLOX]
`RuntimeError: CUDA error: out of memory`

- tests/integration/cli/detection/test_tiling_instseg.py::TestToolsTilingInstanceSegmentation::test_otx_hpo[Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50]
`RuntimeError: CUDA error: CUBLAS_STATUS_ALLOC_FAILED when calling 'cublasCreate(handle)'`

**Comment by yunchu** [comment]
> There are 5 failed cases in pre-merge test. It seems that they are not related to Self-SL for semantic segmentation.
> 
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_otx_demo[Custom_Image_Classification_EfficientNet-V2-S]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficientNet-V2-S]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficinetNet-B0]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/detection/test_detection.py::TestToolsMPADetection::test_otx_explain[Custom_Object_Detection_YOLOX]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/detection/test_tiling_instseg.py::TestToolsTilingInstanceSegmentation::test_otx_hpo[Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50]
>   `RuntimeError: CUDA error: CUBLAS_STATUS_ALLOC_FAILED when calling 'cublasCreate(handle)'`

@sungchul2  FYI, On current CI machine, RTX 2080TI (32G) is installed.

**Comment by sungchul1** [comment]
> There are 5 failed cases in pre-merge test. It seems that they are not related to Self-SL for semantic segmentation.
> 
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_otx_demo[Custom_Image_Classification_EfficientNet-V2-S]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficientNet-V2-S]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification::test_nncf_eval[Custom_Image_Classification_EfficinetNet-B0]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/detection/test_detection.py::TestToolsMPADetection::test_otx_explain[Custom_Object_Detection_YOLOX]
>   `RuntimeError: CUDA error: out of memory`
> * tests/integration/cli/detection/test_tiling_instseg.py::TestToolsTilingInstanceSegmentation::test_otx_hpo[Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50]
>   `RuntimeError: CUDA error: CUBLAS_STATUS_ALLOC_FAILED when calling 'cublasCreate(handle)'`

I conducted local tests about the above failed cases.
- tests/integration/cli/classification/test_classification.py::TestToolsMPAClassification
image

- tests/integration/cli/detection/test_detection.py::TestToolsMPADetection
image

- tests/integration/cli/detection/test_tiling_instseg.py::TestToolsTilingInstanceSegmentation
image

**Review by JihwanEom — APPROVED** [review]

**Pull request merged**
